plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java

```diff
@@ -162,12 +172,31 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
     {
         HudiTableHandle handle = (HudiTableHandle) tableHandle;
         HudiPredicates predicates = HudiPredicates.from(constraint.getSummary());
+        TupleDomain<HiveColumnHandle> regularColumnPredicates = predicates.getRegularColumnPredicates();
+        TupleDomain<HiveColumnHandle> partitionColumnPredicates = predicates.getPartitionColumnPredicates();
+        Set<HiveColumnHandle> constraintColumns = Stream.concat(
+                        Stream.concat(
+                                regularColumnPredicates.getDomains().stream()
+                                        .map(Map::keySet)
+                                        .flatMap(Collection::stream),
+                                partitionColumnPredicates.getDomains().stream()
+                                        .map(Map::keySet)
+                                        .flatMap(Collection::stream)),
+                        // When there is some predicate which could not be translated into tuple domain,
+                        // such as in cases with complex conditions like 'id = 1 OR name = 'INDIA'.
+                        constraint.getPredicateColumns().stream()
+                                .flatMap(Collection::stream)
+                                .map(HiveColumnHandle.class::cast))
```

Comment:
Does Hudi uses the `Constraint#predicate` ? Like if a query has a complex expression like `id = 1 OR name = 'INDIA'` the connector could access the expression via `Constraint#predicate`i.e Connector could pass in the partition values and check the result of the expression - so that it could either pick a partition or reject. I don't see the usages of the `Constraint#predicate` here - If the predicate is not used - then let's not add those column to `constraintColumns` as they are not participating in the partition pushdown. WDYT ?

cc: @codope I'm not a hudi expert - Please correct me if I miss something

Reply:
I compared the behavior of `hive` and `hudi` when `query_partition_filter_required` is enabled. The partition column used in the query satisfies `query_partition_filter_required` requirement even when the column doesn't contribute to the pushdown.

`Hive`:

```
id --> Non-Part column
ds --> Part column
Query --> "EXPLAIN SELECT * FROM test_required_partition_filter WHERE id = 1 OR ds ='INDIA'"

MaterializedResult{rows=[[Trino version: testversion
Fragment 0 [SOURCE]
    Output layout: [id, a, b, ds]
    Output partitioning: SINGLE []
    Output[columnNames = [id, a, b, ds]]
    │   Layout: [id:integer, a:varchar, b:varchar, ds:varchar]
    │   Estimates: {rows: 1 (22B), cpu: 0, memory: 0B, network: 0B}
    └─ ScanFilter[table = hive:tpch:test_required_partition_filter, filterPredicate = (("id" = 1) OR ("ds" = VARCHAR 'INDIA'))]
           Layout: [id:integer, a:varchar, b:varchar, ds:varchar]
           Estimates: {rows: 1 (22B), cpu: 22, memory: 0B, network: 0B}/{rows: 1 (22B), cpu: 22, memory: 0B, network: 0B}
           a := a:string:REGULAR
           b := b:string:REGULAR
           id := id:int:REGULAR
           ds := ds:string:PARTITION_KEY
               :: [[1]]

]], types=[varchar(768)], setSessionProperties={}, resetSessionProperties=[]}
```

`Hudi`:

```
id --> Non-Part column
dt --> Part column
Query --> "EXPLAIN SELECT name FROM " + HUDI_COW_PT_TBL + " WHERE id = 1 OR dt = '2021-12-09'"


MaterializedResult{rows=[[Trino version: testversion
Fragment 0 [SOURCE]
    Output layout: [name]
    Output partitioning: SINGLE []
    Output[columnNames = [name]]
    │   Layout: [name:varchar]
    │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
    └─ ScanFilterProject[table = hudi:tests.hudi_cow_pt_tbl, filterPredicate = (("id" = BIGINT '1') OR ("dt" = VARCHAR '2021-12-09'))]
           Layout: [name:varchar]
           Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}/{rows: ? (?), cpu: ?, memory: 0B, network: 0B}/{rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           dt := dt:string:PARTITION_KEY
           name := name:string:REGULAR
           id := id:bigint:REGULAR

]], types=[varchar(686)], setSessionProperties={}, resetSessionProperties=[]}
```

Reply:
@Praveen2112 Yes you were right `Constraint#predicate` is not getting utilized during split generation that means all the data files will be read for such cases where predicate does not get translated to `TupleDomain`. In this case `constraint#predicateColumns` does not need to be part of `constraintColumns`.